### PR TITLE
Add mapToList & indexedMapToList Functions

### PR DESCRIPTION
**Pull request by prikhi**
These functions traverse the Array, transform each element, & collect
the results into a List.

This is useful when building HTML from an Array:
```elm
type Exercise
renderExercise : Exercise -> Html msg

renderExercises : Array Exercise -> List (Html msg)
renderExercises =
    mapToList renderExercise
```

### CODE REVIEW

**Comment by prikhi** [comment]
@jonboiser Mind if I merge this & cut a new release?

**Comment by jonboiser** [comment]
@prikhi That's fine with me. Was this function requested in an issue? How will it be useful?

**Comment by prikhi** [comment]
> Was this function requested in an issue?

Nah, just going through my projects and pulling out functions I find myself writing all the time.

> How will it be useful?

I use it when rendering an array of items into HTML:
```elm
type SomeItem = ...

renderItem : SomeItem -> Html msg

itemList : Array SomeItem -> Html msg
itemList items =
    Html.ul [] <| Array.mapToList renderItem items
```

That way I don't do a double pass over all the items like with `Array.toList >> List.map renderingFunction` or `Array.map renderingFunction >> Array.toList`.

File: src/Array/Extra.elm [hunk 1]
```diff
@@ -137,6 +137,27 @@ apply fs xs =
...
+    mapToList Html.text : Array String -> List (Html msg)
+
+-}
+mapToList : (a -> b) -> Array a -> List b
+mapToList f =
+    Array.foldl (f >> (::)) []
+
+
+{-| Apply a function to the array with the index as the first argument,
+collecting the results in a List.
```
Comment (jonboiser) [inline src/Array/Extra.elm]: Could you add an example for this function too?

**Comment by jonboiser** [comment]
Sounds good.

I'm having a little trouble with the names, though. I had to read the code to understand that `mapToList` meant something like 'map (over an array) then convert to a list'. But initially I read it as something like 'convert a map to a list'.

Do you have any other ideas for better names.

**Comment by prikhi** [comment]
> initially I read it as something like 'convert a map to a list'

Ah, I assumed it'd be evident since I usually do something like this:

```elm
import Array exposing (Array)
import Array.Extra as Array

renderThem : Array Something -> Html msg
renderThem = div [] <| Array.mapToList (\s -> text s.thing)
```

> Do you have any other ideas for better names

`buildList`, `foldIntoList`, `foldToList`, `toListWith`?

**Comment by jonboiser** [comment]
I don't have any suggestions.

I think this PR is good to merge as soon as you add an example in the docstring for indexedMapToList

**Comment by prikhi** [comment]
added the example, fixed the generated list being in the opposite order of the array, & rebased/squashed the commits.

might make an additional PR to add some tests.

**Comment by jonboiser** [comment]
@prikhi Sounds good! Could you publish after this is merged.

**Pull request merged**